modules/contact/src/actions/ContactAction.C

```diff
@@ -1028,6 +1071,129 @@ ContactAction::addNodeFaceContact()
+  // _mesh->buildBndElemList();
+  const auto & bnd_to_elem_map = _mesh->getBoundariesToActiveSemiLocalElemIds();
+
+  for (const auto & sideset_name : _sideset_list)
+  {
+    // If the sideset provided in the input file isn't in the mesh, error out.
+    const auto find_set = sideset_ids.find(_mesh->getBoundaryID(sideset_name));
+    if (find_set == sideset_ids.end())
+      paramError("sideset_list",
+                 "The sideset ",
+                 sideset_name,
+                 " provided for the contact action is not in the mesh. "
+                 "Please revise your input file.");
+
+    auto dofs_set = bnd_to_elem_map.find(_mesh->getBoundaryID(sideset_name));
+
+    // Initialize data for sideset
+    Point center_of_gravity(0, 0, 0);
+    Real accumulated_sideset_area(0);
+
+    // Pointer to lower-dimensional element on the sideset
+    std::unique_ptr<const Elem> side_ptr;
+    const std::unordered_set<dof_id_type> & bnd_elems = dofs_set->second;
+
+    for (auto elem_id : bnd_elems)
+    {
+      const Elem * elem = _mesh->elemPtr(elem_id);
+      unsigned int side = _mesh->sideWithBoundaryID(elem, _mesh->getBoundaryID(sideset_name));
+
+      // update side_ptr
+      elem->side_ptr(side_ptr, side);
+
+      // area of the (linearized) side
+      const auto side_area = side_ptr->volume();
+
+      // position of the side
+      const auto side_position = side_ptr->true_centroid();
+
+      center_of_gravity += side_position * side_area;
+      accumulated_sideset_area += side_area;
+    }
+
+    // Average each element's center of gravity with its area
+    center_of_gravity /= accumulated_sideset_area;
+
+    // Add sideset-cog pair to vector
+    sideset_list_cog.emplace_back(sideset_name, center_of_gravity);
+  }
+
+  // Vectors of distances for each pair
+  std::vector<std::pair<std::pair<BoundaryName, BoundaryName>, Real>> pairs_distances;
+
+  // Assign distances to identify nearby pairs.
+  for (std::size_t i = 0; i < sideset_list_cog.size() - 1; i++)
+    for (std::size_t j = i + 1; j < sideset_list_cog.size(); j++)
+    {
+      const Point & distance_vector = sideset_list_cog[i].second - sideset_list_cog[j].second;
+
+      if (sideset_list_cog[i].first != sideset_list_cog[j].first)
+      {
+        const Real distance = distance_vector.norm();
+        const std::pair pair = std::make_pair(sideset_list_cog[i].first, sideset_list_cog[j].first);
+        pairs_distances.emplace_back(std::make_pair(pair, distance));
+      }
+    }
+
+  mooseInfo("Reference distance between bodies for automatically assigning contact pairs in "
+            "contact action is: ",
+            getParam<Real>("automatic_pair_distance"));
+
+  // Loop over all pairs and add those whose distance isn't trivial
+  // and less than the threshold distance
+  std::vector<std::pair<std::pair<BoundaryName, BoundaryName>, Real>> lean_pairs_distances;
+  for (const auto & pair_distance : pairs_distances)
+    if (pair_distance.second > TOLERANCE &&
```

Comment:
I'm not sure why we need the "> tolerance" check. Also, maybe use <= on the pair distance (although it probably doesn't really change much)? Finally, it seems a little weird to repeatedly get that parameter in a loop like this. We're probably not going to have enough sidesets for this to really be a performance issue, but it seems like it would still be better to get it once before the loop.